### PR TITLE
Sort list of templates, serviceOfferings, diskOfferings etc in the deploy VM wizard

### PR DESCRIPTION
**Pull request by smlshn**
## Description
Adds functionality to sort the data that is available on each in the deploy VM wizard's step by their suitable fields i.e
- templates by name
- serviceOfferings by cpuspeed
- diskOfferings by disksize
- affinityGroups by name
- sshkeyPairs by name
- network:
   - vpcs by name
   - networkOfferings by name

Fixes: https://github.com/apache/cloudstack/issues/3050

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

- sorted compute offerings
Screen Shot 2019-05-18 at 02 24 53

- sorted disk offerings
Screen Shot 2019-05-18 at 02 24 58

- sorted networks
Screen Shot 2019-05-18 at 02 25 09

## How Has This Been Tested?
Run client
Instances > Add Instance
See previous non sorted data in the browser's console
After sort function run see sorted data in the browser's console
See sorted results in the deploy VM wizard as well

### CODE REVIEW

**Comment by ustcweizhou** [comment]
LGTM

**Comment by smlshn** [comment]
> This overrides the order in which we set the "sort_key" param is used by backend to return a pre-configured list of entities. That is not desirable.
> 
> For templates we don't support sort_key on backend yet and hence we need to sort alphabetically for the time being.
> 
> CC/ @rhtyd , @rafaelweingartner , @borisstoyanov

Actually I think that might be more efficient however in Templates view table there is a column that "Order". Order function does not work right now but if I add sort functionality which sort every result by name, than this might be misleading. One may think it is done in purpose.

Beside, the change is not only about sorting the templates in Create VM screen. diskOfferings, vmtemplates etc all need to be sorted as well.
Screen Shot 2019-05-26 at 22 00 30

**Comment by rhtyd** [comment]
Some resources have a sort order defined, it's not right to enforce general alphabetical order.

**Comment by rhtyd** [comment]
@anuragaw can you advise which resources are sortable from backend (i..e have a sort order), for rest this PR can be used.

**Comment by anuragaw** [comment]
Lost track of it and will update the list of resources before end of day tomorrow. Alternately one can search for "sortKey" API Param in `updateXXX` API calls to find an exhaustive list of resources that support this custom sorting.

**Comment by anuragaw** [comment]
I am afraid most of the entities targetted in the PR description have their own sortkey fields now in code after some recent PRs EXCEPT below-
1. affinityGroups
2. ISOs
3. sshkeyPairs by name
4. network: vpcs by name & networkOfferings by name

**Comment by rhtyd** [comment]
Ping @anuragaw @shwstppr can you review and test this. Thanks.

File: ui/scripts/ui-custom/instanceWizard.js [hunk 1]
```diff
@@ -979,6 +988,8 @@
 
                                     $vpcSelect.val(-1);
 
+                                    sortArrayByKey(args.data.networkOfferings, 'name');
```
Comment (anuragaw) [inline ui/scripts/ui-custom/instanceWizard.js]: Network offerings also has a support for sort_key param in database. We should not sort this.

Comment (anuragaw) [reply ui/scripts/ui-custom/instanceWizard.js]: I had missed this earlier in my list. Apologies for that.

**Pull request merged**